https://www.sqlitetutorial.net/sqlite-python/creating-database/#:~:text=When%20you%20connect%20to%20an,function%20of%20the%20sqlite3%20module.

In [1]:
# Создадим Базу данных sales.db
import sqlite3
from sqlite3 import Error

def create_connection(db_file):
    """ create a database connection to a SQLite database """
    conn = None
    try:
        conn = sqlite3.connect(db_file)
        print(sqlite3.version)
    except Error as e:
        print(e)
    finally:
        if conn:
            conn.close()


if __name__ == '__main__':
    import os
    create_connection(os.path.join("sales.db"))

2.6.0


**Создание таблиц справочников и основной таблицы**

https://habr.com/ru/articles/754400/

In [2]:
# Устанавливаем соединение с базой данных
connection = sqlite3.connect(os.path.join("sales.db"))
cursor = connection.cursor()

# Создаем таблицу Branch
cursor.execute('''
CREATE TABLE IF NOT EXISTS Branch(
Branch TEXT PRIMARY KEY
)
''')

# Создаем таблицу City
cursor.execute('''
CREATE TABLE IF NOT EXISTS City(
City TEXT PRIMARY KEY
)
''')

# Создаем таблицу Product_Line
cursor.execute('''
CREATE TABLE IF NOT EXISTS Product_line(
Product_Line TEXT PRIMARY KEY
)
''')

# Создаем таблицу Customer_Type
cursor.execute('''
CREATE TABLE IF NOT EXISTS Customer_type(
Customer_Type TEXT PRIMARY KEY
)
''')

# Создаем таблицу Gender
cursor.execute('''
CREATE TABLE IF NOT EXISTS Gender(
Gender TEXT PRIMARY KEY
)
''')

# Создаем таблицу Payment
cursor.execute('''
CREATE TABLE IF NOT EXISTS Payment(
Payment TEXT PRIMARY KEY
)
''')


# Создадим таблицу Sales
cursor.execute('''
CREATE TABLE IF NOT EXISTS Sales(
Invoice_ID TEXT,
Branch TEXT, 
City TEXT,
Customer_Type TEXT,
Gender TEXT,
Product_Line TEXT,
Unit_Price REAL,
Quantity INTEGER,
Total REAL,
Date TEXT,
Time TEXT,
Payment REAL,
Payments_fee REAL,
Gross_income REAL,
Cost_price REAL,
Unit_cost REAL
)
''')

Заполнение справочников и БД

In [9]:
!pip install pandas

  Using cached pandas-2.1.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (18 kB)
  Using cached numpy-1.26.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (61 kB)
  Using cached pytz-2023.3.post1-py2.py3-none-any.whl.metadata (22 kB)
  Using cached tzdata-2023.3-py2.py3-none-any.whl (341 kB)
Using cached pandas-2.1.3-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (12.3 MB)
Using cached numpy-1.26.2-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (18.2 MB)
Using cached pytz-2023.3.post1-py2.py3-none-any.whl (502 kB)


In [3]:
import os
import pandas as pd
df = pd.read_csv(os.path.join("clear_data.csv"))
df

Invoice_ID Branch       City Customer_type  Gender  \
0     765-26-6951      A     Yangon        Normal    Male   
1     530-90-9855      A     Yangon        Member    Male   
2     891-01-7034      B   Mandalay        Normal    Male   
3     493-65-6248      C  Naypyitaw        Member  Female   
4     556-97-7101      C  Naypyitaw        Normal  Female   
...           ...    ...        ...           ...     ...   
1164  291-55-6563      A     Yangon        Member  Female   
1165  361-85-2571      A     Yangon        Normal  Female   
1166  364-34-2972      C  Naypyitaw        Member    Male   
1167  115-38-7382      C  Naypyitaw        Member  Female   
1168  115-38-7382      C  Naypyitaw        Member  Female   

                Product_line  Unit_price  Quantity   Total  \
0          Sports and travel       76.24         6  457.44   
1         Home and lifestyle       49.97         8  399.76   
2     Electronic accessories       78.45         6  470.67   
3          Sports and travel       38.83        10  388.29   
4     Electronic accessories       66.38         2  132.76   
...                      ...         ...       ...     ...   
1164      Home and lifestyle       36.14         6  216.85   
1165       Sports and travel       93.95         5  469.77   
1166  Electronic accessories      101.66         3  304.98   
1167     Fashion accessories        4.64         8   37.12   
1168     Fashion accessories        4.97         8   39.75   

                               Date                           Time  \
0     2019-01-01 00:00:00.000000000  2023-11-05 10:39:00.000000000   
1     2019-01-01 00:00:00.000000000  2023-11-05 14:47:00.000000000   
2     2019-01-01 00:00:00.000000000  2023-11-05 19:07:00.000000000   
3     2019-01-01 00:00:00.000000000  2023-11-05 19:48:00.000000000   
4     2019-01-01 00:00:00.000000000  2023-11-05 15:51:00.000000000   
...                             ...                            ...   
1164  2019-03-30 00:00:00.000000000  2023-11-05 12:45:00.000000000   
1165  2019-03-30 00:00:00.000000000  2023-11-05 10:18:00.000000000   
1166  2019-03-30 00:00:00.000000000  2023-11-05 20:37:00.000000000   
1167  2019-03-30 00:00:00.000000000  2023-11-05 12:51:00.000000000   
1168  2019-03-30 00:00:00.000000000  2023-11-05 12:51:00.000000000   

          Payment  Payments_fee  Gross_income  Cost_price  Unit_cost  
0     Credit card         13.72         68.62      388.82      64.80  
1            Cash          0.00         71.96      327.80      40.98  
2            Cash          0.00         70.60      400.07      66.68  
3     Credit card         11.65         58.24      330.05      33.01  
4            Cash          0.00         19.91      112.85      56.43  
...           ...           ...           ...         ...        ...  
1164      Ewallet          3.25         39.03      177.82      29.64  
1165         Cash          0.00         70.47      399.30      79.86  
1166         Cash          0.00         45.75      259.23      86.41  
1167  Credit card          1.11          9.28       27.84       3.48  
1168  Credit card          1.19          9.94       29.81       3.73  

[1169 rows x 16 columns]

In [81]:
# поставим необходимую для загрузки данных библиотеку
!pip install SQLAlchemy

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 6.0 MB/s eta 0:00:0000:0100:01
Using cached greenlet-3.0.1-cp310-cp310-manylinux_2_24_x86_64.manylinux_2_28_x86_64.whl (613 kB)


In [4]:
# Добавим в БД справочники:
# https://www.alixaprodev.com/how-to-insert-pandas-dataframe-to-sqlite-table-in-python/
# https://stackoverflow.com/questions/53189071/insert-a-pandas-dataframe-into-a-sqlite-table-update-a-table-with-a-dataframe

from sqlalchemy import create_engine
engine = create_engine('sqlite:///sales.db', echo=False)

# ["Branch", "City", "Product_Line", "Customer_Type", "Customer_Type", "Gender", "Payment"]:
for cl in ["Branch", "City", "Product_line", "Customer_type", "Gender", "Payment"]:
    print(cl)
    temp_df = df[[cl]]
    temp_df.drop_duplicates(inplace=True)
    print(temp_df)
    temp_df.to_sql(cl, con=engine, if_exists='replace', index=False)

    

Branch
  Branch
0      A
2      B
3      C


/tmp/ipykernel_45488/3635520179.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temp_df.drop_duplicates(inplace=True)
/tmp/ipykernel_45488/3635520179.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temp_df.drop_duplicates(inplace=True)


City
        City
0     Yangon
2   Mandalay
3  Naypyitaw
Product_line
               Product_line
0         Sports and travel
1        Home and lifestyle
2    Electronic accessories
5         Health and beauty
6       Fashion accessories
10       Food and beverages
12  Clothes and accessories


/tmp/ipykernel_45488/3635520179.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temp_df.drop_duplicates(inplace=True)


Customer_type
  Customer_type
0        Normal
1        Member
Gender
   Gender
0    Male
3  Female


/tmp/ipykernel_45488/3635520179.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temp_df.drop_duplicates(inplace=True)
/tmp/ipykernel_45488/3635520179.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temp_df.drop_duplicates(inplace=True)


Payment
       Payment
0  Credit card
1         Cash
7      Ewallet


/tmp/ipykernel_45488/3635520179.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  temp_df.drop_duplicates(inplace=True)


In [5]:
# Добавим продажи
df.to_sql("Sales", con=engine, if_exists='replace', index=False)

1169

## Исследование данных с помощью библиотеки sqlite3

**1. В каких городах были продажи**

In [15]:
connection = sqlite3.connect(os.path.join("sales.db"))
cursor = connection.cursor()

# ожидается такой ответ:
query = "SELECT DISTINCT(City) FROM Sales"
rows = cursor.execute(query)
print([x[0] for x in rows])

['Yangon', 'Mandalay', 'Naypyitaw']


In [16]:
# второй вариант, тривиальный, но не противоречащий заданию:
query = "SELECT * FROM City"
rows = cursor.execute(query)
print([x[0] for x in rows])

['Yangon', 'Mandalay', 'Naypyitaw']


**2. Какая категория товара лидирует по числу продаж в ед. товара, а какая по суммарным продажам?**

In [20]:
query = "SELECT Product_line, SUM(Quantity), SUM(Total) FROM Sales GROUP BY Product_line"
rows = cursor.execute(query)
for x in rows:
    print(x)

('Clothes and accessories', 829, 18188.40000000001)
('Electronic accessories', 971, 54337.63999999999)
('Fashion accessories', 1015, 56769.979999999996)
('Food and beverages', 952, 56144.96000000001)
('Health and beauty', 854, 49193.84000000002)
('Home and lifestyle', 911, 53861.960000000036)
('Sports and travel', 920, 55123.00000000001)


Из запроса видно, что 
- по числу продаж лидирует Fashion accessories, продано 1015 единиц
- по выручке также лидирует Fashion accessories, продано на сумму 56769.98



In [21]:
# Можно сделать красивее, превратив полученные списки кортежей в датафрейм (но задание этого не требует):
query = "SELECT Product_line, SUM(Quantity), SUM(Total) FROM Sales GROUP BY Product_line"
rows = cursor.execute(query)
rows_df = pd.DataFrame(rows, columns=["Product_line", "Quantity", "Total"])
rows_df

Product_line  Quantity     Total
0  Clothes and accessories       829  18188.40
1   Electronic accessories       971  54337.64
2      Fashion accessories      1015  56769.98
3       Food and beverages       952  56144.96
4        Health and beauty       854  49193.84
5       Home and lifestyle       911  53861.96
6        Sports and travel       920  55123.00

**3. В какой день были минимальные, а в какой максимальные продажи по суммарной выручке за день?**

In [30]:
# Выведем день с максимальными продажами
query_max = '''SELECT Date, max(DaySale) FROM 
            (SELECT Date, SUM(Total) as DaySale FROM Sales 
                GROUP BY Date)'''
rows = cursor.execute(query_max)
list(rows)

[('2019-03-09 00:00:00.000000000', 8164.920000000001)]

In [37]:
# Выведем день с минимальными продажами
query_min = '''SELECT Date, min(DaySale) FROM 
            (SELECT Date, SUM(Total) as DaySale FROM Sales 
                GROUP BY Date)'''
rows = cursor.execute(query_min)
list(rows)

[('2019-02-13 00:00:00.000000000', 1172.78)]

**Замечание**

Полученная минимальные сумма продаж и дата формально правильная, но вдумчивый аналитик, должен догадаться, что в данных какая-то ошибка (мы сами ее внесли, когда усредняли пропущенные значения). Надо заменить эту дату на '2019-02-13 00:00:00.000000000' и повторить поиск минимальной суммы и даты

In [36]:
query_update = """UPDATE Sales SET Date = '2019-02-13 00:00:00.000000000'
                    WHERE Date = '2019-02-13 23:51:22.191780864'"""
cursor.execute(query_update)

In [38]:
# Еще раз получим день с минимальными продажами 
rows = cursor.execute(query_min)
list(rows)

[('2019-02-13 00:00:00.000000000', 1172.78)]

Таким образом, 
- максимальный продажи были 9 марта и составили 8164.92
- минимальные продажи были 13 февраля и составили 1172.78


**4. Женщины или мужчины приносят больше выручки?**

In [41]:
query = "SELECT Gender, SUM(Total) FROM Sales GROUP BY Gender"
rows = cursor.execute(query)
list(rows)

[('Female', 193316.2500000001), ('Male', 150303.52999999997)]

Видим, что больше выручки принесли женщины 193316.25 vs 150303.53 у мужчин 

**5. Отличаются ли мужской средний чек от женского?**

In [48]:
# Посчитаем средний чек у женщин

query = """SELECT Gender, COUNT(*), SUM(Total), SUM(Total)/COUNT(*) as AverageBill  
            FROM Sales GROUP BY Gender"""
rows = cursor.execute(query)
list(rows)

[('Female', 636, 193316.2500000001, 303.9563679245284),
 ('Male', 533, 150303.52999999997, 281.9953658536585)]

Средний чек у женщин = 303,96 больше, чем у мужчин = 282

**6. Вывести выручку и маржинальную прибыль в разрезе товарных категорий**

In [49]:
query = "SELECT Product_line, SUM(Gross_income), SUM(Total) FROM Sales GROUP BY Product_line"
rows = cursor.execute(query)
rows_df = pd.DataFrame(rows, columns=["Product_line", "Gross_income", "Total"])
rows_df

Product_line      Quantity     Total
0  Clothes and accessories   1273.300000  18188.40
1   Electronic accessories   8150.720000  54337.64
2      Fashion accessories  14242.092877  56769.98
3       Food and beverages   6737.400000  56144.96
4        Health and beauty  10822.700000  49193.84
5       Home and lifestyle   9695.140000  53861.96
6        Sports and travel   8268.480000  55123.00

In [50]:
# Из любви к искусству добавим маржинальную рентабельность:
query = "SELECT Product_line, SUM(Gross_income), SUM(Total), SUM(Gross_income)/SUM(Total) FROM Sales GROUP BY Product_line"
rows = cursor.execute(query)
rows_df = pd.DataFrame(rows, columns=["Product_line", "Gross_income", "Total", "Profitability"])
rows_df

Product_line  Gross_income     Total  Profitability
0  Clothes and accessories   1273.300000  18188.40       0.070006
1   Electronic accessories   8150.720000  54337.64       0.150001
2      Fashion accessories  14242.092877  56769.98       0.250874
3       Food and beverages   6737.400000  56144.96       0.120000
4        Health and beauty  10822.700000  49193.84       0.220001
5       Home and lifestyle   9695.140000  53861.96       0.180000
6        Sports and travel   8268.480000  55123.00       0.150001